src/tools/kata-ctl/src/ops/volume_ops.rs

```diff
@@ -0,0 +1,212 @@
+// Copyright (c) 2022 Boston University
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+
+use crate::args::{DirectVolSubcommand, DirectVolumeCommand};
+
+use anyhow::{anyhow, Ok, Result};
+use futures::executor;
+use hyper;
+use kata_types::mount::DirectVolumeMountInfo;
+use std::{fs, path::Path, path::PathBuf, time::Duration};
+use url;
+
+use runtimes;
+
+const TIMEOUT: Duration = Duration::new(5, 0);
+const CONTENT_TYPE_JSON: &str = "application/json";
+const MOUNT_INFO_FILE_NAME: &str = "mountInfo.json";
+const KATA_DIRECT_VOLUME_ROOT_PATH: &str = "/run/kata-containers/shared/direct-volumes";
+
+pub fn handle_direct_volume(vol_cmd: DirectVolumeCommand) -> Result<()> {
+    let command = vol_cmd.directvol_cmd;
+    match command {
+        DirectVolSubcommand::Add(args) => {
+            add(&args.volume_path, &args.mount_info)?;
+        }
+        DirectVolSubcommand::Remove(args) => {
+            remove(&args.volume_path)?;
+        }
+        DirectVolSubcommand::Stats(args) => {
+            let result = executor::block_on(stats(&args.volume_path))?;
+            println!("{}", &result);
+        }
+        DirectVolSubcommand::Resize(args) => {
+            executor::block_on(resize(&args.volume_path, args.resize_size))?;
+        }
+    }
+    Ok(())
+}
+
+async fn resize(volume_path: &str, size: u64) -> Result<()> {
+    let sandbox_id = get_sandbox_id_for_volume(volume_path)?;
+    let mount_info = get_volumn_mount_info(volume_path)?;
+    let resize_req = runtimes::ResizeRequest {
+        size,
+        volume_path: mount_info.device,
+    };
+    let encoded = serde_json::to_string(&resize_req)?;
+    let shim_client = runtimes::MgmtClient::new(sandbox_id, Some(TIMEOUT)).unwrap();
+
+    let url = runtimes::DIRECT_VOLUMN_RESIZE_URL;
+    let response = shim_client
+        .post(url, &String::from(CONTENT_TYPE_JSON), &encoded)
+        .await?;
+    if response.status() != hyper::StatusCode::OK {
+        return Err(anyhow!(
+            "status {}: failed to resize volume",
+            response.status()
+        ));
+    }
+    Ok(())
+}
+
+async fn stats(volume_path: &str) -> Result<String> {
+    let sandbox_id = get_sandbox_id_for_volume(volume_path)?;
+    let mount_info = get_volumn_mount_info(volume_path)?;
+
+    let req_url =
+        url::form_urlencoded::Serializer::new(String::from(runtimes::DIRECT_VOLUMN_STATS_URL))
+            .append_pair(runtimes::DIRECT_VOLUMN_PATH_KEY, &mount_info.device)
+            .finish();
+
+    // sid: sandbox id
+    let shim_client = runtimes::MgmtClient::new(sandbox_id, Some(TIMEOUT)).unwrap();
+    shim_client.get(&req_url).await?;
+    let response = shim_client.get(&req_url).await?;
+    // turn body into string
+    let body = hyper::body::to_bytes(response.into_body()).await?;
+    let body = String::from_utf8(body.to_vec()).unwrap();
+    Ok(body)
+}
+
+// join_path joins user provided volumepath with kata direct-volume root path
+fn join_path(volume_path: &str) -> PathBuf {
```

Comment:
- What if `volume_path == ""`?
- Should `volume_path` be an absolute path (starting with `/`)? This would be preferred from a security perspective, but if the value should be absolute, this function should return `Result<PathBuf>` so it can detect bad paths like `../../foo`.

Adding table driven tests are great for shaking out and handling corner cases and invalid values like this: https://github.com/kata-containers/kata-containers/blob/main/docs/Unit-Test-Advice.md#rust-table-driven-tests

Reply:
The join_path function in `add` and `remove` encode `volume_path` into b64 string [also in kata-runtime](https://github.com/kata-containers/kata-containers/blob/cd27ad144e1a111cb606015c5c9671431535e644/src/runtime/pkg/direct-volume/utils.go#L59) so I think some of the mentioned issues can be mitigated, however since b64 permits `/` character, `safe-path::scoped_join`  is indeed necessary😀